File: src/main/java/org/wysko/midis2jam2/instrument/pipe/BlownBottle.java
package org.wysko.midis2jam2.instrument.pipe;

import com.jme3.math.FastMath;
import com.jme3.math.Quaternion;
import com.jme3.scene.Node;
import com.jme3.scene.Spatial;
import org.wysko.midis2jam2.Midis2jam2;
import org.wysko.midis2jam2.instrument.brass.WrappedOctaveSustained;
import org.wysko.midis2jam2.midi.MidiChannelSpecificEvent;
import org.wysko.midis2jam2.particle.SteamPuffer;

import java.util.List;
import java.util.stream.IntStream;

import static org.wysko.midis2jam2.Midis2jam2.MatType.REFLECTIVE;
import static org.wysko.midis2jam2.Midis2jam2.rad;

public class BlownBottle extends WrappedOctaveSustained {
	Node[] bottleNodes = new Node[12];
	
	public BlownBottle(Midis2jam2 context, List<MidiChannelSpecificEvent> events) {
		super(context, events, true);
		
		twelfths = new Bottle[12];
		
		IntStream.range(0, 12).forEach(i -> bottleNodes[i] = new Node());
		
		for (int i = 0; i < 12; i++) {
			twelfths[i] = new Bottle(i);
			twelfths[i].highestLevel.setLocalTranslation(-15, 0, 0);
			
			bottleNodes[i].attachChild(twelfths[i].highestLevel);
			bottleNodes[i].setLocalRotation(new Quaternion().fromAngles(0, rad(7.5 * i), 0));
			bottleNodes[i].setLocalTranslation(0, 0.3f * i, 0);
			instrumentNode.attachChild(bottleNodes[i]);
		}
		
		
		instrumentNode.setLocalTranslation(75, 0, -35);
	}
	
	@Override
	protected void moveForMultiChannel() {
		offsetNode.setLocalTranslation(0, 20 + indexForMoving() * 3.6f, 0);
		instrumentNode.setLocalRotation(new Quaternion().fromAngles(0, FastMath.HALF_PI * indexForMoving(), 0));
	}
	
	public class Bottle extends WrappedOctaveSustained.TwelfthOfOctave {
		SteamPuffer puffer;
		
		public Bottle(int i) {
			this.puffer = new SteamPuffer(context, SteamPuffer.SteamPuffType.POP, 1);
			highestLevel.attachChild(context.loadModel("PopBottle.obj", "PopBottle.bmp", REFLECTIVE, 0.9f));
			Spatial label = context.loadModel("PopBottleLabel.obj", "PopLabel.bmp");
			label.setLocalRotation(new Quaternion().fromAngles(0, FastMath.PI, 0));
			highestLevel.attachChild(label);
			Spatial pop = context.loadModel("PopBottlePop.obj", "Pop.bmp", REFLECTIVE, 0.8f);
			Spatial middle = context.loadModel("PopBottleMiddle.obj", "PopBottle.bmp", REFLECTIVE,
					0.9f);
			float scale = 0.3f + 0.027273f * i;
			pop.setLocalTranslation(0, -3.25f, 0);
			pop.scale(1, scale, 1);
			middle.scale(1, 1 - scale, 1);
			highestLevel.attachChild(pop);
			highestLevel.attachChild(middle);
			highestLevel.attachChild(puffer.steamPuffNode);
			puffer.steamPuffNode.setLocalRotation(new Quaternion().fromAngles(0, FastMath.PI, 0));
			puffer.steamPuffNode.setLocalTranslation(1, 3.5f, 0);
		}
		
		@Override
		public void play(double duration) {
			playing = true;
			progress = 0;
			this.duration = duration;
		}
		
		@Override
		public void tick(double time, float delta) {
			if (progress >= 1) {
				playing = false;
				progress = 0;
			}
			if (playing) {
				progress += delta / duration;
			}
			puffer.tick(time, delta, playing);
		}
	}
}

File: app/src/commonMain/kotlin/org/wysko/midis2jam2/instrument/family/pipe/BlownBottle.kt
/*
 * Copyright (C) 2025 Jacob Wysko
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see https://www.gnu.org/licenses/.
 */
package org.wysko.midis2jam2.instrument.family.pipe

import org.wysko.kmidi.midi.TimedArc
import org.wysko.kmidi.midi.event.MidiEvent
import org.wysko.midis2jam2.Midis2jam2
import org.wysko.midis2jam2.instrument.DivisiveSustainedInstrument
import org.wysko.midis2jam2.instrument.PitchClassAnimator
import org.wysko.midis2jam2.instrument.algorithmic.PitchBendModulationController
import org.wysko.midis2jam2.midi.notePeriodsModulus
import org.wysko.midis2jam2.particle.SteamPuffer
import org.wysko.midis2jam2.particle.SteamPuffer.Behavior.Outwards
import org.wysko.midis2jam2.particle.SteamPuffer.Texture.Pop
import org.wysko.midis2jam2.util.*
import org.wysko.midis2jam2.world.modelD
import org.wysko.midis2jam2.world.modelR
import kotlin.time.Duration

/** The Blown bottle. */
class BlownBottle(context: Midis2jam2, events: List<MidiEvent>) :
    DivisiveSustainedInstrument(context, events) {

    override val animators: List<PitchClassAnimator> = List(12) { Bottle(it, events.notePeriodsModulus(context, it)) }

    private val pitchBendModulationController = PitchBendModulationController(context, events, 0.0)

    init {
        repeat(12) {
            with(geometry) {
                +node {
                    +animators[it].root.also {
                        it.loc = v3(-15, 0, 0)
                    }
                    loc = v3(0, 0.3 * it, 0)
                    rot = v3(0, 7.5 * it, 0)
                }
            }
        }
        placement.loc = v3(75, 0, -35)
    }

    override fun adjustForMultipleInstances(delta: Duration) {
        with(updateInstrumentIndex(delta)) {
            root.loc = v3(0, 20 + this * 3.6, 0)
            placement.rot = v3(0, 90 * this, 0)
        }
    }

    override fun tick(time: Duration, delta: Duration) {
        super.tick(time, delta)
        pitchBendModulationController.tick(time, delta)
    }

    /** A single Bottle. */
    inner class Bottle(i: Int, notePeriods: List<TimedArc>) : PitchClassAnimator(context, notePeriods) {

        private val puffer: SteamPuffer = SteamPuffer(context, Pop, 1.0, Outwards)
        private val bendCtrl = NumberSmoother(0f, 10.0)

        override fun tick(time: Duration, delta: Duration) {
            super.tick(time, delta)
            puffer.tick(delta, playing)
            animation.loc = v3(
                x = 0,
                y = bendCtrl.tick(delta) { if (playing) pitchBendModulationController.bend else 0f },
                z = 0
            )
        }

        init {
            // Load pop bottle
            with(geometry) {
                +context.modelR("PopBottle.obj", "PopBottle.bmp")
                +context.modelD("PopBottleLabel.obj", "PopLabel.bmp").apply { rot = v3(0, 180, 0) }

                val scale = 0.3f + 0.027273f * i
                +context.modelR("PopBottlePop.obj", "Pop.bmp").apply {
                    loc = v3(0, -3.25, 0)
                    scale(1f, scale, 1f)
                }
                +context.modelR("PopBottleMiddle.obj", "PopBottle.bmp").apply {
                    scale(1f, 1 - scale, 1f)
                }

                +puffer.root.apply {
                    loc = v3(1, 3.5, 0)
                    rot = v3(0, 180, 0)
                }
            }
        }
    }
}
